From 586d5e9dbd88411c7af7bbc3012cd1b06a26a957 Mon Sep 17 00:00:00 2001
From: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
Date: Tue, 3 Feb 2026 12:06:18 +0200
Subject: [PATCH 182/187] [WIP-YUV420] drm/display: hdmi_state_helper: Allow
 fallback after explicit color format

This is particularly helpful when color format has been explicitly set
to YUV420, since this is normally supported only by a (very) limited
number of display modes (at least according to the information
advertised by sink's EDID).

Without the fallback, all those modes that do not support YUV420 will
fail, unless the color format is once again explicitly set to something
else, e.g. RGB.

Hence interpret the color format as a preferred option rather than an
enforced one.

Signed-off-by: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
---
 .../gpu/drm/display/drm_hdmi_state_helper.c   | 43 +++++++++++--------
 1 file changed, 24 insertions(+), 19 deletions(-)

diff --git a/drivers/gpu/drm/display/drm_hdmi_state_helper.c b/drivers/gpu/drm/display/drm_hdmi_state_helper.c
index ceabfdb30f7a..029919354da7 100644
--- a/drivers/gpu/drm/display/drm_hdmi_state_helper.c
+++ b/drivers/gpu/drm/display/drm_hdmi_state_helper.c
@@ -655,27 +655,32 @@ hdmi_compute_config(const struct drm_connector *connector,
 
 	ret = hdmi_compute_format_bpc(connector, conn_state, mode, max_bpc,
 				      hdmi_colorspace);
-	if (ret) {
-		/* If a color format was explicitly requested, don't fall back */
-		if (conn_state->color_format) {
-			drm_dbg_kms(connector->dev,
-				    "Explicitly set color format '%s' doesn't work.\n",
-				    drm_get_color_format_name(conn_state->color_format));
-			return ret;
-		}
+	if (!ret)
+		return ret;
+
+	if (conn_state->color_format)
+		drm_dbg_kms(connector->dev,
+			    "Explicitly set color format '%s' doesn't work.\n",
+			    drm_get_color_format_name(conn_state->color_format));
 
-		if (connector->ycbcr_420_allowed) {
-			ret = hdmi_compute_format_bpc(connector, conn_state,
-						      mode, max_bpc,
-						      HDMI_COLORSPACE_YUV420);
-			if (ret)
-				drm_dbg_kms(connector->dev,
-					    "YUV420 output format doesn't work.\n");
-		} else {
+	if (hdmi_colorspace != HDMI_COLORSPACE_RGB)
+		ret = hdmi_compute_format_bpc(connector, conn_state,
+					      mode, max_bpc,
+					      HDMI_COLORSPACE_RGB);
+	if (!ret || hdmi_colorspace == HDMI_COLORSPACE_YUV420)
+		return ret;
+
+	if (connector->ycbcr_420_allowed) {
+		ret = hdmi_compute_format_bpc(connector, conn_state,
+					      mode, max_bpc,
+					      HDMI_COLORSPACE_YUV420);
+		if (ret)
 			drm_dbg_kms(connector->dev,
-				    "YUV420 output format not allowed for connector.\n");
-			ret = -EINVAL;
-		}
+				    "YUV420 output format doesn't work.\n");
+	} else {
+		drm_dbg_kms(connector->dev,
+			    "YUV420 output format not allowed for connector.\n");
+		ret = -EINVAL;
 	}
 
 	return ret;
-- 
2.43.0

